File: sdks/php/src/Service/MarketingEmails.php
<?php

namespace Kibamail\Service;

use Kibamail\Collection;
use Kibamail\Enums\Transporter\ContentType;
use Kibamail\Enums\Transporter\Method;
use Kibamail\MarketingEmail;
use Kibamail\Resource;
use Kibamail\ValueObjects\ResourceUri;
use Kibamail\ValueObjects\Transporter\Payload;

class MarketingEmails extends Service
{
    /**
     * Create a marketing email.
     *
     * @param array{
     *     name: string,
     *     subject?: string,
     *     emailContent?: array,
     *     from?: string,
     *     replyTo?: string
     * } $parameters Marketing email creation parameters
     *
     * @return MarketingEmail The created marketing email
     */
    public function create(array $parameters): MarketingEmail
    {
        $payload = Payload::create('v1/marketing-emails', $parameters);
        $result = $this->transporter->request($payload);

        return $this->createResource('marketing-emails', $result);
    }

    /**
     * List all marketing emails.
     *
     * @param array{'limit'?: int, 'before'?: string, 'after'?: string} $options
     */
    public function list(array $options = []): Collection
    {
        $payload = Payload::list('v1/marketing-emails', $options);
        $result = $this->transporter->request($payload);

        return $this->createResource('marketing-emails', $result);
    }

    /**
     * Retrieve a single marketing email by ID.
     */
    public function get(string $id): MarketingEmail
    {
        $payload = Payload::get('v1/marketing-emails', $id);
        $result = $this->transporter->request($payload);

        return $this->createResource('marketing-emails', $result);
    }

    /**
     * Update a marketing email by ID.
     *
     * @param string $id The marketing email ID
     * @param array{
     *     name?: string,
     *     subject?: string,
     *     emailContent?: array,
     *     from?: string,
     *     replyTo?: string
     * } $parameters Marketing email update parameters
     *
     * @return MarketingEmail The updated marketing email
     */
    public function update(string $id, array $parameters): MarketingEmail
    {
        $payload = Payload::update('v1/marketing-emails', $id, $parameters);
        $result = $this->transporter->request($payload);

        return $this->createResource('marketing-emails', $result);
    }

    /**
     * Delete a marketing email by ID.
     */
    public function delete(string $id): MarketingEmail
    {
        $payload = Payload::delete('v1/marketing-emails', $id);
        $result = $this->transporter->request($payload);

        return $this->createResource('marketing-emails', $result);
    }

    /**
     * Preview a marketing email by ID.
     */
    public function preview(string $id): Resource
    {
        $payload = new Payload(
            ContentType::JSON,
            Method::GET,
            ResourceUri::withAction('v1/marketing-emails', $id, 'preview')
        );
        $result = $this->transporter->request($payload);

        return $this->createResource('marketing-emails', $result);
    }

    /**
     * Retrieve statistics for a marketing email.
     */
    public function stats(string $id): Resource
    {
        $payload = new Payload(
            ContentType::JSON,
            Method::GET,
            ResourceUri::withAction('v1/marketing-emails', $id, 'stats')
        );
        $result = $this->transporter->request($payload);

        return $this->createResource('marketing-emails', $result);
    }
}

File: sdks/go/cmd/kibamail/cmd/marketing_emails.go
package cmd

import (
	"github.com/kibamail/cli/internal"
	kibamail "github.com/kibamail/kibamail/packages/go-sdk"
	"github.com/spf13/cobra"
)

func newMarketingEmailsCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "marketing-emails",
		Short: "Manage marketing emails",
	}
	cmd.AddCommand(
		newMarketingEmailsCreateCmd(),
		newMarketingEmailsListCmd(),
		newMarketingEmailsShowCmd(),
		newMarketingEmailsUpdateCmd(),
		newMarketingEmailsDeleteCmd(),
		newMarketingEmailsPreviewCmd(),
		newMarketingEmailsStatsCmd(),
	)
	return cmd
}

func newMarketingEmailsCreateCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "create",
		Short: "Create a marketing email",
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireClient(cmd); err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			name, _ := cmd.Flags().GetString("name")
			req := &kibamail.CreateMarketingEmailRequest{Name: name}
			if v, _ := cmd.Flags().GetString("subject"); v != "" {
				req.Subject = v
			}
			if v, _ := cmd.Flags().GetString("preview-text"); v != "" {
				req.PreviewText = v
			}
			if v, _ := cmd.Flags().GetString("html"); v != "" {
				req.Html = v
			}
			if v, _ := cmd.Flags().GetString("sender-identity-id"); v != "" {
				req.SenderIdentityId = v
			}
			if v, _ := cmd.Flags().GetString("reply-to-identity-id"); v != "" {
				req.ReplyToIdentityId = v
			}
			if v, _ := cmd.Flags().GetString("type"); v != "" {
				req.Type = v
			}
			if cmd.Flags().Changed("track-clicks") {
				v, _ := cmd.Flags().GetBool("track-clicks")
				req.TrackClicks = &v
			}
			if cmd.Flags().Changed("track-opens") {
				v, _ := cmd.Flags().GetBool("track-opens")
				req.TrackOpens = &v
			}
			result, err := Client.MarketingEmails.Create(req)
			if err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			if internal.IsJSON(cmd) {
				return internal.PrintResult(cmd, result)
			}
			cmd.Printf("Created marketing email %s\n", result.ID)
			return nil
		},
	}
	cmd.Flags().String("name", "", "Marketing email name [required]")
	cmd.Flags().String("subject", "", "Email subject")
	cmd.Flags().String("preview-text", "", "Preview text")
	cmd.Flags().String("html", "", "HTML body")
	cmd.Flags().String("sender-identity-id", "", "Sender identity ID")
	cmd.Flags().String("reply-to-identity-id", "", "Reply-to identity ID")
	cmd.Flags().String("type", "", "Email type")
	cmd.Flags().Bool("track-clicks", false, "Track clicks")
	cmd.Flags().Bool("track-opens", false, "Track opens")
	_ = cmd.MarkFlagRequired("name")
	return cmd
}

func newMarketingEmailsListCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "list",
		Short: "List marketing emails",
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireClient(cmd); err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			opts := &kibamail.ListOptions{}
			if limit, _ := cmd.Flags().GetInt("limit"); limit > 0 {
				opts.Limit = &limit
			}
			if after, _ := cmd.Flags().GetString("after"); after != "" {
				opts.After = &after
			}
			result, err := Client.MarketingEmails.List(opts)
			if err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			return internal.PrintResult(cmd, result)
		},
	}
	cmd.Flags().Int("limit", 0, "Maximum number of results")
	cmd.Flags().String("after", "", "Cursor for next page")
	return cmd
}

func newMarketingEmailsShowCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "show <id>",
		Short: "Show a marketing email",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireClient(cmd); err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			result, err := Client.MarketingEmails.Get(args[0])
			if err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			return internal.PrintResult(cmd, result)
		},
	}
}

func newMarketingEmailsUpdateCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "update <id>",
		Short: "Update a marketing email",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireClient(cmd); err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			req := &kibamail.UpdateMarketingEmailRequest{}
			if v, _ := cmd.Flags().GetString("name"); v != "" {
				req.Name = v
			}
			if v, _ := cmd.Flags().GetString("subject"); v != "" {
				req.Subject = v
			}
			if v, _ := cmd.Flags().GetString("preview-text"); v != "" {
				req.PreviewText = v
			}
			if v, _ := cmd.Flags().GetString("html"); v != "" {
				req.Html = v
			}
			if v, _ := cmd.Flags().GetString("sender-identity-id"); v != "" {
				req.SenderIdentityId = v
			}
			if v, _ := cmd.Flags().GetString("reply-to-identity-id"); v != "" {
				req.ReplyToIdentityId = v
			}
			if v, _ := cmd.Flags().GetString("type"); v != "" {
				req.Type = v
			}
			if cmd.Flags().Changed("track-clicks") {
				v, _ := cmd.Flags().GetBool("track-clicks")
				req.TrackClicks = &v
			}
			if cmd.Flags().Changed("track-opens") {
				v, _ := cmd.Flags().GetBool("track-opens")
				req.TrackOpens = &v
			}
			result, err := Client.MarketingEmails.Update(args[0], req)
			if err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			if internal.IsJSON(cmd) {
				return internal.PrintResult(cmd, result)
			}
			cmd.Printf("Updated marketing email %s\n", result.ID)
			return nil
		},
	}
	cmd.Flags().String("name", "", "Marketing email name")
	cmd.Flags().String("subject", "", "Email subject")
	cmd.Flags().String("preview-text", "", "Preview text")
	cmd.Flags().String("html", "", "HTML body")
	cmd.Flags().String("sender-identity-id", "", "Sender identity ID")
	cmd.Flags().String("reply-to-identity-id", "", "Reply-to identity ID")
	cmd.Flags().String("type", "", "Email type")
	cmd.Flags().Bool("track-clicks", false, "Track clicks")
	cmd.Flags().Bool("track-opens", false, "Track opens")
	return cmd
}

func newMarketingEmailsDeleteCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "delete <id>",
		Short: "Delete a marketing email",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireClient(cmd); err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			err := Client.MarketingEmails.Delete(args[0])
			if err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			if internal.IsJSON(cmd) {
				cmd.Printf(`{"deleted":true,"id":"%s"}`+"\n", args[0])
			} else {
				cmd.Printf("Deleted marketing email %s\n", args[0])
			}
			return nil
		},
	}
}

func newMarketingEmailsPreviewCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "preview <id>",
		Short: "Preview a marketing email",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireClient(cmd); err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			result, err := Client.MarketingEmails.Preview(args[0])
			if err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			return internal.PrintResult(cmd, result)
		},
	}
}

func newMarketingEmailsStatsCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "stats <id>",
		Short: "Show statistics for a marketing email",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireClient(cmd); err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			result, err := Client.MarketingEmails.Stats(args[0])
			if err != nil {
				internal.HandleError(cmd, err)
				return nil
			}
			return internal.PrintResult(cmd, result)
		},
	}
}
